clippy_lints/src/use_shared_from_slice.rs

```diff
@@ -0,0 +1,135 @@
+use rustc::hir;
+use rustc::hir::{Expr, Local};
+use rustc::lint::{LateContext, LateLintPass, LintArray, LintPass};
+use rustc::ty;
+use crate::utils::{match_qpath, match_type, match_type_parameter, snippet, span_lint_and_sugg, walk_ptrs_ty, get_type_parameter};
+use crate::utils::paths;
+
+/// **What it does:**
+/// Checks for usage of `Rc<String>` or `Rc<Vec<T>>`.
+///
+/// **Why is this bad?**
+/// Using `Rc<str>` or `Rc<[T]>` is more efficient and easy to construct with
+/// `into()`.
+///
+/// **Known problems:**
+/// None.
+///
+/// **Example:**
+///
+/// ```rust
+/// use std::rc::Rc;
+///
+/// // Bad
+/// let bad_ref: Rc<Vec<usize>> = Rc::new(vec!(1, 2, 3));
+///
+/// // Good
+/// let good_ref: Rc<[usize]> = vec!(1, 2, 3).into();
+/// ```
+declare_clippy_lint! {
+    pub USE_SHARED_FROM_SLICE,
+    nursery,
+    "constructing reference-counted type from `Vec` or `String`"
+}
+
+#[derive(Copy, Clone)]
+pub struct Pass;
+
+impl LintPass for Pass {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(USE_SHARED_FROM_SLICE)
+    }
+}
+
+/// If the given `expr` is constructing an `Rc` or `Arc` containing a `Vec` or
+/// `String`, output a suggestion to fix accordingly.
+fn check_rc_construction<'a, 'tcx>(cx: &LateContext<'a, 'tcx>, expr: &'tcx Expr) {
+    let expr_ty = walk_ptrs_ty(cx.tables.expr_ty(expr));
+
+    // Check for expressions with the type `Rc<Vec<T>>`.
+    if_chain! {
+        // Check if this expression is constructing an `Rc` or `Arc`.
+        let is_rc = match_type(cx, expr_ty, &paths::RC);
+        let is_arc = match_type(cx, expr_ty, &paths::ARC);
+        if is_rc || is_arc;
+
+        // Check if the `Rc` or `Arc` is constructed with `Vec` or `String`.
+        if let ty::TyAdt(_, subst) = expr_ty.sty;
+        let arg_type = subst.type_at(0);
+        let arg_is_vec = match_type(cx, arg_type, &paths::VEC);
+        let arg_is_string = match_type(cx, arg_type, &paths::STRING);
+        if arg_is_vec || arg_is_string;
+
+        // Get the argument, to use for writing out the lint message.
+        if let hir::ExprCall(_, ref args) = expr.node;
+        if let Some(arg) = args.get(0);
+
+        then {
+            if arg_is_vec {
```

Comment:
So this is a little dangerous, because the message might appear even if the user has no way to change the type. For `Rc<Vec<[T]>>` this is is kinda fine, because the suggestion will still work, but for `Rc<String>` I'm not sure if `.as_str().into()` is equivalent.